### PR TITLE
ddl: modify different character sets varchar type column maximum length limit

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 

We execute the following sql , found that tidb and mysql behave differently.
```sql
# tidb master branch / tidb v2.1.0:

tidb>  CREATE TABLE tr (id INT, name VARCHAR(20000), purchased DATE )  DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
ERROR 1074 (42000): Column length too big for column 'name' (max = 16383); use BLOB or TEXT instead
```
```sql
# tidb v2.0.0 :

MySQL [test]> CREATE TABLE tr (id INT, name VARCHAR(20000), purchased DATE )  DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.01 sec)

mysql>  CREATE TABLE tr (id INT, name VARCHAR(65536), purchased DATE )  DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
ERROR 1074 (42000): Column length too big for column 'name' (max = 21845); use BLOB or TEXT instead
```

```sql
# mysql 5.7
mysql>  CREATE TABLE tr (id INT, name VARCHAR(65536), purchased DATE )  DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
ERROR 1074 (42000): Column length too big for column 'name' (max = 16383); use BLOB or TEXT instead

mysql>  CREATE TABLE tr (id INT, name VARCHAR(65536), purchased DATE )  DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
ERROR 1074 (42000): Column length too big for column 'name' (max = 21845); use BLOB or TEXT instead

mysql>  CREATE TABLE tr (id INT, name VARCHAR(65536), purchased DATE ) DEFAULT CHARSET=latin1;
ERROR 1074 (42000): Column length too big for column 'name' (max = 65535); use BLOB or TEXT instead
```
### What is changed and how it works?
* Different character sets varchar type column maximum length limit is not the same.

* [There are default settings for character sets and collations at four levels: server, database, table, and column.](https://dev.mysql.com/doc/refman/5.7/en/charset-syntax.html)

1. First use the character set of Column to set the value;
2. If not set above, use the default character set value of the corresponding table;
3. If not set above, use the default character set value of the corresponding database;
3. If not set above, use the character_set_server setting.

### Check List 

Tests 

 - Unit test
 

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8295)

### CODE REVIEW

**Comment by ciscoxll** [comment]
@winkyao @zimulala @crazycs520 PTAL

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -657,6 +657,30 @@ func checkColumnsAttributes(colDefs []*ast.ColumnDef) error {
...
 
+func checkTooBigFieldLength(schema *model.DBInfo, colDefs []*ast.ColumnDef, tbInfo *model.TableInfo) error {
+	for _, colDef := range colDefs {
+		if colDef.Tp.Tp == mysql.TypeVarchar {
+			cs := tbInfo.Charset
+			if len(tbInfo.Charset) == 0 {
+				if len(schema.Charset) == 0 {
+					cs = mysql.DefaultCharset
+				}
+				cs = schema.Charset
```
Comment (kennytm) [inline ddl/ddl_api.go]: Wouldn't this always overwrite line 666 then?

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -491,11 +491,10 @@ func checkColumn(colDef *ast.ColumnDef) error {
 	case mysql.TypeVarchar:
 		maxFlen := mysql.MaxFieldVarCharLength
 		cs := tp.Charset
-		// TODO: TableDefaultCharset-->DatabaseDefaultCharset-->SystemDefaultCharset.
 		// TODO: Change TableOption parser to parse collate.
 		// Reference https://github.com/pingcap/tidb/blob/b091e828cfa1d506b014345fb8337e424a4ab905/ddl/ddl_api.go#L185-L204
 		if len(tp.Charset) == 0 {
-			cs = mysql.DefaultCharset
+			return nil
```
Comment (winkyao) [inline planner/core/preprocess.go]: Add comments to explain why return nil here.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -657,6 +657,31 @@ func checkColumnsAttributes(colDefs []*ast.ColumnDef) error {
...
+					cs = mysql.DefaultCharset
+				} else {
+					cs = schema.Charset
+				}
+			}
+			desc, err := charset.GetCharsetDesc(cs)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			maxFlen := mysql.MaxFieldVarCharLength
```
Comment (winkyao) [inline ddl/ddl_api.go]: Use common function to extract the logic, don't write redundant code.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -989,6 +1014,15 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	if err = checkTooBigFieldLength(schema, s.Cols, tbInfo); err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(tbInfo.Charset) != 0 && strings.ToLower(tbInfo.Charset) != mysql.DefaultCharset {
```
Comment (winkyao) [inline ddl/ddl_api.go]: Use `IsUTF8Charset`

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -989,6 +1014,15 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
...
+		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf(`TiDB only supports the "utf8mb4" character set, so "%s" does not take effect.`, tbInfo.Charset))
```
Comment (winkyao) [inline ddl/ddl_api.go]: No need to append warning here.

File: ddl/ddl_api.go (hunk 3 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: Done.

**Comment by ciscoxll** [comment]
@winkyao @zimulala @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@winkyao @zimulala @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@winkyao @zimulala @crazycs520 PTAL

**Comment by ciscoxll** [comment]
@zimulala @crazycs520 PTAL

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -658,6 +658,42 @@ func checkColumnsAttributes(colDefs []*ast.ColumnDef) error {
...
+		if colDef.Tp.Tp == mysql.TypeVarchar {
+			var setCharset string
+			setCharset = mysql.DefaultCharset
+			if len(schema.Charset) != 0 {
+				setCharset = schema.Charset
+			}
+			if len(tbInfo.Charset) != 0 {
+				setCharset = tbInfo.Charset
+			}
+
```
Comment (crazycs520) [inline ddl/ddl_api.go]: Please add some comment to explain that `colDef.Tp.Charset` will be check in `planner/core/preprocess.go`, So ignore here.

Comment (crazycs520) [reply ddl/ddl_api.go]: I got some problem here:
In mysql: 
```sql
mysql root@127.0.0.1:test> create table t1 (a varchar(20000) ) default charset utf8;
Query OK, 0 rows affected
Time: 0.008s
mysql root@127.0.0.1:test> show create table t1
+-------+--------------------------------------+
| Table | Create Table                         |
+-------+--------------------------------------+
| t1    | CREATE TABLE `t1` (                  |
|       |   `a` varchar(20000) DEFAULT NULL    |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
+-------+--------------------------------------+
1 row in set
Time: 0.003s
mysql root@127.0.0.1:test> alter table t1 convert to character set utf8mb4;
(1074, u"Column length too big for column 'a' (max = 16383); use BLOB or TEXT instead")
```

In TiDB This PR:
```sql
mysql root@127.0.0.1:test> create table t1 (a varchar(20000) ) default charset utf8;
Query OK, 0 rows affected
Time: 0.005s
mysql root@127.0.0.1:test> alter table t1 convert to character set utf8mb4;
Query OK, 0 rows affected
Time: 0.005s
mysql root@127.0.0.1:test> show create table t1
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                         |
|       |   `a` varchar(20000) DEFAULT NULL                           |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
```

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -658,6 +658,42 @@ func checkColumnsAttributes(colDefs []*ast.ColumnDef) error {
 	return nil
 }
 
+// checkColumnFieldLength check the maximum length limit for different character set varchar type columns.
+func checkColumnFieldLength(schema *model.DBInfo, colDefs []*ast.ColumnDef, tbInfo *model.TableInfo) error {
+	for _, colDef := range colDefs {
+		if colDef.Tp.Tp == mysql.TypeVarchar {
+			var setCharset string
```
Comment (winkyao) [inline ddl/ddl_api.go]: should we modify the tbInfo.Charset if it is empty?

File: planner/core/preprocess.go [hunk 8]
```diff
@@ -489,22 +488,16 @@ func checkColumn(colDef *ast.ColumnDef) error {
...
-		cs := tp.Charset
-		// TODO: TableDefaultCharset-->DatabaseDefaultCharset-->SystemDefaultCharset.
-		// TODO: Change TableOption parser to parse collate.
-		// Reference https://github.com/pingcap/tidb/blob/b091e828cfa1d506b014345fb8337e424a4ab905/ddl/ddl_api.go#L185-L204
+		// Here logical move backend processing, need to get all levels of character set to set the character set of column.
+		// It's not easy to get the schema charset and table charset here.
+		// ColumnDefaultCharset --> TableDefaultCharset-->DatabaseDefaultCharset-->SystemDefaultCharset.
 		if len(tp.Charset) == 0 {
-			cs = mysql.DefaultCharset
+			return nil
```
Comment (winkyao) [inline planner/core/preprocess.go]: // It's not easy to get the schema charset and table charset here. The charset is determined by the order `ColumnDefaultCharset --> TableDefaultCharset-->DatabaseDefaultCharset-->SystemDefaultCharset`.
// return nil, to make the check in the ddl.CreateTable.

Comment (ciscoxll) [reply planner/core/preprocess.go]: Done.

File: ddl/ddl_api.go (hunk 7 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: If `tbInfo.Charset` is empty, tidb will not modify it.

**Pull request merged**